Fix clock placement in devices with limited grid

rw1nkler opened the pull request:
This commit adds the ability to verify if the clock location is available in the FPGA grid before placement. It uses architecture XML for the verification.

Resolves https://github.com/SymbiFlow/symbiflow-arch-defs/issues/1517

mithro commented:
PTAL - @tcal-x

tcal-x commented:
> PTAL - @tcal-x

I will try the script on the test where I had to add the workaround LOC constraints.   I need to rebuild rr_graph in my sandbox though so it will be a few hours before I have an update.

tcal-x commented:
This fix may be unrelated to the issue I'm seeing.  In any case, I see exactly the same error using the new version of the script, just the line numbers are different.   In my case, the script has difficulty placing the PLLE2_ADV if it is not constrained.  I'm targeting the full 100T chip.

This is the error (line numbers are using the new version from this PR):
```
Traceback (most recent call last):
  File "/home/tcal/2nd-tcal-x/symbiflow-arch-defs/xc/common/utils/prjxray_create_place_constraints.py", line 734, in <module>
    main()
  File "/home/tcal/2nd-tcal-x/symbiflow-arch-defs/xc/common/utils/prjxray_create_place_constraints.py", line 722, in main
    vpr_grid, loc_in_use, block_locs, blocks, grid_capacities):
  File "/home/tcal/2nd-tcal-x/symbiflow-arch-defs/xc/common/utils/prjxray_create_place_constraints.py", line 519, in place_clocks
    for potential_loc in sorted(available_placements[key]):
KeyError: ('PLLE2_ADV', 5)
```

rw1nkler commented:
I have changed the script to use limited device grid. Additionally, now it is possible to use only top/bottom half of the device. 

The script does not check if the used resource is available in the given clock domain. (This bug was present before) Because of that, after limiting the device grid, sometimes it tries to place i.e. `PLLE2_ADV` instance in the clock domain that, does not have it. I fixed the problem by excluding almost empty clock domains from the `basys-bottom` device definition when the problem occurs.

A universal solution requires rearranging the constraint problem and adding proper checks of variable domains to clocking resources like i.e. PLLs.

litghost commented on xc/common/utils/prjxray_create_place_constraints.py:
```diff
@@ -200,15 +210,21 @@ def __init__(self, vpr_grid, io_locs, blif_data, roi):
+            _, thresh_bot_y = cmt_dict[bot_cmt_tile]['canon_loc']
+        except StopIteration:
+            thresh_bot_y = None
 
```
Should likely assert that either thresh_bot_y or thresh_top_y is present.  Otherwise things will likely break horribly.

litghost replied:
Also if graph limit is not set, both thresh_bot_y and thresh_top_y should be `not None`.

rw1nkler replied:
I have added the requested changes

litghost commented on xc/common/utils/prjxray_create_place_constraints.py:
```diff
@@ -189,26 +199,39 @@ def get_tile_dict(self):
+            assert thresh_bot_y is not None, "CMTs in the bottom half of the device not found"
+        else:
+            assert thresh_top_y is not None or thresh_bot_y is not None, (
+                "The device grid does not contain any CMTs"
```
A better message would be "does not contain any BUFG sites".  If now BUFG's are available, then any source to BUFG connections become effectively impossible, without something special.

rw1nkler replied:
Should I also change the previous assert messages to "BUFG sites in the top/bottom half of the device not found"?

litghost replied:
Ya, that is more precise.

litghost commented on xc/common/utils/prjxray_create_place_constraints.py:
```diff
@@ -189,26 +199,39 @@ def get_tile_dict(self):
 
 class ClockPlacer(object):
-    def __init__(self, vpr_grid, io_locs, blif_data, roi):
+    def __init__(self, vpr_grid, io_locs, blif_data, args):
```
Rather than pass args directly, please pass roi and clock limit explicitly.  We want to avoid tying ClockPlacer directly to argparse.

rw1nkler replied:
Fixed
